plugins/out_s3/s3.c

```diff
@@ -960,9 +985,18 @@ static int s3_put_object(struct flb_s3 *ctx, const char *tag, time_t create_time
         c = mock_s3_call("TEST_PUT_OBJECT_ERROR", "PutObject");
     }
     else {
-        c = s3_client->client_vtable->request(s3_client, FLB_HTTP_PUT,
-                                              uri, body, body_size,
-                                              NULL, 0);
+        h = create_canned_acl_header(ctx->canned_acl);
+        if (h == NULL) {
```

Comment:
continue with my comment on your create function, I think it'd be more typical do this logic as:
```
if canned_acl == NULL {
normal case
} 
else {
create header
make call with header
}

```

Reply:
its about separating responsibility. Your create_canned_acl_header has a simple job- just create the header. The caller has the job of deciding whether or not it needs to create the header

Reply:
so the logic should like this way?
`if canned_acl == NULL {
normal case
} 
else {
create header
    if header == NULL {
    same as canned_acl == NULL} 
    else {
    }
}`

Reply:
if the header returned by `create_canned_acl_header` is NULL then that should mean it failed to create the header, which means you are in an error case. Then you do not make the API call and instead should do error handling

Reply:
> if header == NULL {
    same as canned_acl == NULL
     } 
My comment refers to that part

Reply:
this is a standard practice- returning NULL means something went wrong

Reply:
I see. Thanks.